### ISSUE TITLE
Client fetching subtitles takes too long

### ISSUE THREAD

**Issue by vitorsemeano** [jellyfin/jellyfin#935]
**Describe the bug**
When a jellyfin client tries to get a specific subtitle from a video file that contains subtitle streams, it takes some minutes to get a response from the server. The requested link must be in the following format: 

http://<server address>/emby/Videos/<video id>/<video id>/Subtitles/3/0/Stream.the subtitle format like vtt or ass>?api_key=<api key>

**To Reproduce**
To reproduce, you must have a video file with embed subtitles (subtitle streams, not hardcoded). Make sure you didn't played the file with subtitles yet. Try to restart the emby service to clean cached info (not sure how it caches the parsed subtitles, so it's a wild guess).

1. Open the webserver client or android app
2. Navigate to a file that meets the video file requirements
3. For the subtitle dropdown, select one subtitle
4. Click to play it
5. It can happen two things:
- The file plays instantly, but without subtitles
- The loading circle keeps spinning for some minutes

**Expected behavior**
After point 4, the video file should play normally with the pretended subtitles.

**Logs**
Can't see any error logs related with this issue, the request made by the client side simply takes a lot of time for a first time. Seeing the logs in jellyfin server only gives the request received and returned a few minutes later.

**Screenshots**

**System (please complete the following information):**
 - OS: ubuntu 18.10 64bit. The clients used was the jellyfin webserver included with the server, and the android app built with this version: https://github.com/vitorsemeano/jellyfin-android/tree/fixSubtitles
 - Browser: Firefox
 - Jellyfin Version: 10.1.0

**Additional context**
This issue could be related with #172

**Comment by definitio** [comment]
This is still an issue.

**Comment by Rassilion** [comment]
Happened to me in docker armhf and docker amd64 images.

**Comment by ovandriyanov** [comment]
Happens to me all the time. Kinda annoying. Is there any workaround? E. g. an API call to force subtitle extraction and caching?

**Comment by RazeLighter777** [comment]
Workaround is to deselect and reselect the subtitles

**Comment by Rassilion** [comment]
@razelighter777 that doesn't work for me. Are you using external subs or internal? I'm using internal subs in mkv container.

**Comment by ovandriyanov** [comment]
@RazeLighter777 I can't do it when the video is played with selected subtitles from the very beginning, e. g. when subtitles are enabled by default in settings, or when I select them before playing a video, since in that case I don't have any user interface at all — only the spinning loading circle at the center of my screen.

**Comment by RazeLighter777** [comment]
@Rassilion Internal mkv subtitles.

**Comment by JustAMan** [comment]
@ovandriyanov 
> Happens to me all the time. Kinda annoying. Is there any workaround? E. g. an API call to force subtitle extraction and caching?

There should be a setting in JF admin corner that allows pre-extraction of subtitles (though I might be mixing it up with "Allow on-the-fly extraction"), pls check that.
Also check Jellyfin 10.4.0 as it has some fixes related to how video starts playing when subtitles are concerned.

**Comment by Rassilion** [comment]
@JustAMan 

> There should be a setting in JF admin corner that allows pre-extraction of subtitles (though I might be mixing it up with "Allow on-the-fly extraction"), pls check that.

There is only "Allow on-the-fly extraction", if some kind of pre-extraction option can be added probably fix this issue

> Also check Jellyfin 10.4.0 as it has some fixes related to how video starts playing when subtitles are concerned.

I tested 10.4.0 issue is still there. Subtitles starts like 30-40 seconds after video starts

**Comment by dannymichel** [comment]
I think this is a bug as it is still an issue, and if it's not an issue on plex/emby, it isn't 'normal'

**Comment by dkanada** [comment]
I agree with the first statement but not the second. Someone added a change a while back to load the subtitles asynchronously, which is probably the culprit here. @MrTimscampi would you be interested in making that configurable?

**Comment by Rassilion** [comment]
@dkanada 
> I agree with the first statement but not the second. Someone added a change a while back to load the subtitles asynchronously, which is probably the culprit here. @MrTimscampi would you be interested in making that configurable?

I think "Allow on-the-fly extraction" option does that. If I remember right, when disabled it blocks video untill subtitles loaded.

**Comment by msawatzky75** [comment]
I have tried the "Allow subtitle extraction on the fly" option (both enabled and disabled) and have not noticed a change with the delay in loading subtitles.

**Comment by itshoro** [comment]
Yes, this is still an issue.

I don't mind if there was an option available (let jellyfin use ffmpeg to extract subs so that they can also be loaded faster or something like that), but this is both annoying and just a bad experience.

Also it's not even an universal issue, as far as I can tell it depends on the amount of fonts being referenced in the file. (Subs that don't reference a specific font load instantaneously from my experience)

Also I don't experience a difference when toggling "Allow subtitle extraction on the fly".

**Comment by itshoro** [comment]
From further research it seems that mostly people that have no hardware acceleration available / enabled are experiencing this problem.

The only workaround I've found so far is to manually run these two commands to extract the subtitles and removing BOM.
```
​ffmpeg -txt_format text -i infile.mkv infile.srt
​find . -iname '*.srt' -print -exec sed -i '1s/^/\xef\xbb\xbf/' {} \;
```

It's really annoying. In my opinion there should be another option. Maybe having the option to run a second ffmpeg instance that only runs 
```
ffmpeg -i file -loglevel quiet -f srt -
```
and dumps it into a buffer so that it can be used. But I'm far from experienced in this field, so take this with a grain of salt.

**Comment by BIoob** [comment]
I have hardware acceleration enabled and still have this issue.

**Comment by cvium** [comment]
Extracting subtitles is pure CPU and I/O.

> Allow subtitle extraction on the fly

This option doesn't do anything and never did. No one got around to fixing it.

**Comment by msawatzky75** [comment]
Slightly unrelated: I recently started using the Chromcast client with the same files that would normally load subtitles slowly, and in the Chromcast client the subtitles appear instantly. So this means that Jellyfin is indeed capable of doing this, just its implemented differently for the web client.

**Comment by cvium** [comment]
Chromecast probably burns the subtitles and that's indeed unrelated

**Comment by msawatzky75** [comment]
I was hoping to provide possible insight into a fix, but I guess not.

**Comment by cvium** [comment]
The problem is that subtitles are extracted on the fly, which is limited by CPU and I/O speed because ffmpeg has to read and decode the entire file to extract the stream.

**Comment by msawatzky75** [comment]
But if I copy the stream url and play it in a video player on my machine (not in browser) the sub titles are there immediately too. Are these also burned in? I have my doubts, but I am not very experienced in this area.

**Comment by hbfrank** [comment]
> The problem is that subtitles are extracted on the fly, which is limited by CPU and I/O speed because ffmpeg has to read and decode the entire file to extract the stream.

That's the real truth.

**Comment by dannymichel** [comment]
> > The problem is that subtitles are extracted on the fly, which is limited by CPU and I/O speed because ffmpeg has to read and decode the entire file to extract the stream.
> 
> 
> 
> That's the real truth.

Yes, I believe we are all aware of why it takes so long, but it's possible to speed it up, and I hate to say this but because Plex and Emby can and does do it. It's immediate on plex.
However, with that said i can't be sure if this is limited to ffmpeg or if plex even uses ffmpeg

**Comment by Rassilion** [comment]
Yes subtitle extraction slows subtitles but I think also there are problems in subtitle rendering in browser.

Subtitles slower in browsers, especially old or different platform browsers(like ps4 or TV os) compared to desktop or native clients.

**Comment by Gee1111** [comment]
Emby dont have this problem.. i think emby only extract the subtitles for the frames it needs now... i think thats the key. Extracting whole subs of a movie is a bad idea

**Comment by febriananur** [comment]
while I am using jellyfin media player the subs load normally but while using in the browser its takes time

**Comment by Visne** [comment]
Is this issue superseded by #6672, or are they different issues?
